### PR TITLE
Chameleon

### PR DESCRIPTION
**Pull request by mfaverge**
Add interface to link with Chameleon

### CODE REVIEW

File: src/h_matrix.hpp [hunk 1]
```diff
@@ -551,8 +551,14 @@ template<typename T> class HMatrix : public Tree<HMatrix<T> >, public RecursionM
...
-    assert(i + j * nrChildRow() < this->nrChild());
-    return this->getChild(i + j * nrChildRow());
+    if ( this->nrChild() ) {
+	assert(i + j * nrChildRow() < this->nrChild());
+	return this->getChild(i + j * nrChildRow());
+    }
+    else {
+	assert(i + j * nrChildRow() == 0);
+	return (HMatrix<T>*)this;
+    }
```
Comment (gsylvand) [inline src/h_matrix.hpp]: In which contest do you need this ? can't you do the test in the caller ?

Comment (mfaverge) [reply src/h_matrix.hpp]: I need it if the HMatrix has only one level. Thus you can only get the HMatrix itself as it has no children.

Comment (gsylvand) [reply src/h_matrix.hpp]: Its better to remove this and do the test in the caller (or change the way your cluster tree is created to be sure to have at least 1 child at the root)

File: src/h_matrix.cpp [hunk 2]
```diff
@@ -2155,6 +2155,118 @@ void HMatrix<T>::solve(FullMatrix<T>* b) const {
...
+    /* Upper case */
+    if ( upper  ) {
+	if ( left ) {
+	    if ( notrans ) {
+		/* LUN */
+		solveUpperTriangularLeft( B, unit, false );
+	    }
+	    else {
+		/* LUT */
+		std::cerr << "ERROR: TRSM LUT case is for now missing !!!" << std::endl;
```
Comment (gsylvand) [inline src/h_matrix.cpp]: Replace all std::cerr with HMAT_ASSERT_MSG(0, "blabla");

File: src/c_wrapping.hpp [hunk 3]
```diff
@@ -500,6 +500,14 @@ void read_data(hmat_matrix_t * matrix, hmat_iostream readfunc, void * user_data)
...
 }
 
+// Read a full hmatrix
+template <typename T, template <typename> class E>
+hmat_matrix_t *read( hmat_iostream readfunc, void *data ) {
+    hmat_matrix_t *hmat = read_struct<T,E>( readfunc, data );
+    read_data<T,E>( hmat, readfunc, data );
+    return hmat;
+}
+
```
Comment (gsylvand) [inline src/c_wrapping.hpp]: keep it within chameleon (idem for write)

File: src/h_matrix.hpp (hunk 1 above)
Comment (mfaverge) [reply src/h_matrix.hpp]: Done

File: src/h_matrix.cpp (hunk 2 above)
Comment (mfaverge) [reply src/h_matrix.cpp]: Done

File: src/h_matrix.hpp (hunk 1 above)
Comment (mfaverge) [reply src/h_matrix.hpp]: Done

File: src/c_wrapping.hpp (hunk 3 above)
Comment (mfaverge) [reply src/c_wrapping.hpp]: Done

File: src/h_matrix.cpp [hunk 4]
```diff
@@ -2155,6 +2155,118 @@ void HMatrix<T>::solve(FullMatrix<T>* b) const {
...
```
Comment (gsylvand) [inline src/h_matrix.cpp]: you forgot those 3 !

Comment (mfaverge) [reply src/h_matrix.cpp]: oops :)

**Pull request merged**